sympy/core/basic.py

```diff
@@ -1275,18 +1299,35 @@ def has_free(self, *patterns):
         {y}
         >>> expr.has_free(g(y))
         True
-        >>> expr.has_free(*(x, f(x)))
+        >>> expr.has_free(x, f(x))
         False
 
-        This works for subexpressions and types, too:
+        This works for literal subexpressions and types, too:
 
         >>> expr.has_free(g)
         True
         >>> (x + y + 1).has_free(y + 1)
         True
-
+        >>> (2*x*y + 3).has_free(x*y)
+        True
         """
-        return self._has(iterfreeargs, *patterns)
+        if not patterns:
+            return False
+        p0 = patterns[0]
+        if len(patterns) == 1 and iterable(p0) and not isinstance(p0, Basic):
+            # Basic can contain iterables (though not non-Basic, ideally)
+            # but don't encourage mixed passing patterns
+            raise ValueError('expecting one or more Basic patterns')
+        # try quick test first
+        s = set(patterns)
+        rv = self.has_free_arg(s)
+        if rv:
+            return rv
+        # now try matching through slower _has
+        for a in iterfreeargs(self):
+            if not getattr(a, '_diff_wrt', False) and a._has(iterfreeargs, *s):
```

Comment:
Won't `iterfreeargs(self)` yield all of the things that `iterfreeargs(a)` could yield? This looks like an inefficient loop that will go over subexpressions many more times than needed.

Reply:
good point